### PR TITLE
Mihai Popa's assignment

### PR DESCRIPTION
**Pull request by mihai12**
Removed node_modules and added it to git ignore

### CODE REVIEW

File: coding-challenge/src/bank/validation.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+export class validation {
```
Comment (adrianmota-sag) [inline coding-challenge/src/bank/validation.ts:1]: I like the ideea! But I think we are overcomplicating things a bit with this validation class for this particular case.

File: coding-challenge/tst/bank/bank-account.test.ts [hunk 2]
```diff
@@ -20,7 +20,7 @@ describe('Tests for bank account class', () => {
         const bankAccountBalance = bankAccount.checkBalance();
 
         // Assert
-        expect(bankAccountBalance).toBe(initialBalance);
+        expect(bankAccountBalance).toEqual(initialBalance);
```
Comment (adrianmota-sag) [inline coding-challenge/tst/bank/bank-account.test.ts:23]: Why did you decide to change toBe to toEqual?

File: coding-challenge/tst/bank/bank-account.test.ts [hunk 3]
```diff
@@ -34,30 +34,50 @@ describe('Tests for bank account class', () => {
...
-        }).toThrow('Withdraw amount has to be greater than 0!');
+        }).toThrowError('Withdraw amount has to be greater than 0!');
     });
 
     it('should throw a "Insufficient funds!" error', () => {
         // Act & Assert
         expect(() => {
             bankAccount.withdraw(initialBalance + 1)
-        }).toThrow('Insufficient funds!');
+        }).toThrowError('Insufficient funds!');
```
Comment (adrianmota-sag) [inline coding-challenge/tst/bank/bank-account.test.ts:44]: Why did you change toThrow to toThrowError?

File: coding-challenge/tst/bank/bank-account.test.ts [hunk 4]
```diff
@@ -34,30 +34,50 @@ describe('Tests for bank account class', () => {
...
 
     it('should update the account balance with the specified sum!', () => {
+        // Arange
+        const depositAmount = 100;
+
+        //Act
+        bankAccount.deposit(depositAmount);
 
+        //Assert
+        expect(bankAccount.checkBalance()).toEqual(initialBalance + depositAmount);
```
Comment (adrianmota-sag) [inline coding-challenge/tst/bank/bank-account.test.ts:80]: Which do you think is best here, toBe or toEqual?

File: coding-challenge/tst/bank/bank-account.test.ts [hunk 5]
```diff
@@ -81,13 +101,35 @@ describe('Tests for bank account class', () => {
         bankAccount.transfer(transferAmount, destinationAccount);
 
         // Assert
-        expect(bankAccount['balance']).toBe(initialBalance - transferAmount);
-        expect(destinationAccount['balance']).toBe(transferAmount);
+        expect(bankAccount['balance']).toEqual(initialBalance - transferAmount);
```
Comment (adrianmota-sag) [inline coding-challenge/tst/bank/bank-account.test.ts:104]: Same as above...

File: coding-challenge/tst/bank/bank-account.test.ts [hunk 6]
```diff
@@ -34,30 +34,50 @@ describe('Tests for bank account class', () => {
...
     it('should withdraw the specified amount and update the current balance', () => {
+        // Arrange
+        const withdrawAmount = 45;
 
-    });
+        // Act 
+        bankAccount.withdraw(withdrawAmount);
 
+        // Assert
+        expect(bankAccount.checkBalance()).toEqual(initialBalance - withdrawAmount);
```
Comment (razvan-zavalichi) [inline coding-challenge/tst/bank/bank-account.test.ts:55]: We shouldn't get the `balance` using a public function. If `checkBalance` has some errors this test will fail even if the `withdraw` function is working properly.

File: coding-challenge/tst/bank/bank-account.test.ts [hunk 7]
```diff
@@ -81,13 +101,35 @@ describe('Tests for bank account class', () => {
...
 
+        const depositAction = jest.spyOn(destinationAccount, 'deposit').mockImplementationOnce(() => {
+            throw new Error('Transfer operation failed!');
+        })
+
+        // Act & Assert
+        expect(() => {
+            bankAccount.transfer(transferAmount, destinationAccount)
+        }).toThrowError('Transfer operation failed!');
+        expect(bankAccount.checkBalance()).toEqual(initialBalance);
```
Comment (adrianmota-sag) [inline coding-challenge/tst/bank/bank-account.test.ts:131]: Same as above..

File: coding-challenge/tst/bank/bank-account.test.ts [hunk 8]
```diff
@@ -81,13 +101,35 @@ describe('Tests for bank account class', () => {
...
+        expect(destinationAccount['balance']).toEqual(transferAmount);
     });
 
+    it('throws an error if the balance is insufficient', () => {
+        // Arrange
+        const destinationAccount: BankAccount = new BankAccount('DestinationAccount');
+
+        // Act & Assert
+        expect(() => {
+            bankAccount.transfer(50000, destinationAccount)
```
Comment (razvan-zavalichi) [inline coding-challenge/tst/bank/bank-account.test.ts:114]: `50000` deservers a const.

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 7 above)
Comment (adrianmota-sag) [reply coding-challenge/tst/bank/bank-account.test.ts:131]: 1.methods should not be used to test methods in a unit test
2.toEqual vs toBe?

File: coding-challenge/src/bank/validation.ts (hunk 1 above)
Comment (razvan-zavalichi) [inline coding-challenge/src/bank/validation.ts:1]: This class does not have unit tests.

File: coding-challenge/tst/bank/bank-account.test.ts [hunk 9]
```diff
@@ -81,13 +101,35 @@ describe('Tests for bank account class', () => {
...
+        }).toThrowError('Insufficient funds!');
+      });
+
     it('should transfer the specified sum back to the source account if the deposit operation fails!', () => {
+        // Arrange
+        const destinationAccount: BankAccount = new BankAccount('DestinationAccount');
+        const transferAmount = 100;
 
+        const depositAction = jest.spyOn(destinationAccount, 'deposit').mockImplementationOnce(() => {
```
Comment (adrianmota-sag) [inline coding-challenge/tst/bank/bank-account.test.ts:123]: Is there any reason that you decided to mock de deposit function but not the withdraw function?

File: coding-challenge/src/bank/bank-account.ts [hunk 10]
```diff
@@ -1,32 +1,36 @@
...
         return this.balance;
     };
 
-    transfer(transferAmount: number, destinationBankAccount: BankAccount) {
-      // This method should take a sum out of the source account and transfer it to the destination bank account.
+    public transfer(transferAmount: number, destinationBankAccount: BankAccount): void {
+      validation.validateAmount(transferAmount, "Transfer");
+      validation.validateFunds(this.balance, transferAmount);
+      this.withdraw(transferAmount);
+      destinationBankAccount.deposit(transferAmount);
```
Comment (adrianmota-sag) [inline coding-challenge/src/bank/bank-account.ts:33]: What happens here if the withdraw method is successful but the deposit method fails for some reason?

File: coding-challenge/src/bank/validation.ts (hunk 1 above)
Comment (00noize00) [reply coding-challenge/src/bank/validation.ts:1]: Why is this public? Maybe switch the repo to private... Just sayin, had it in watch and I am seeing all the comments and commits

File: coding-challenge/src/bank/validation.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    public static validateAmount(amount: number, fnc: string): void {
+        if (amount <= 0) {
+          throw new Error(`${fnc} amount has to be greater than 0!`);
```
Comment (adrianmota-sag) [inline coding-challenge/src/bank/validation.ts:4]: Generic functions should not throw errors, this should just return the result of the validation (true/false) and the parent component decides what to do with the response (throw an error, go on a different branch etc...)

File: coding-challenge/src/bank/validation.ts (hunk 1 above)
Comment (mihai12) [reply coding-challenge/src/bank/validation.ts:1]: Initially this was inside the BankAccount class, thought it would make a good impression to show that I can think about utility-like classes for universal use in a project so I’ve hastily moved the validation functions to their own class, made them static, and forgot to remove the throw new error as well, guess I shot myself in the leg here.

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 2 above)
Comment (mihai12) [reply coding-challenge/tst/bank/bank-account.test.ts:23]: I was thinking that the users can input amounts that have decimals, and I know that toBe is not good with floating numbers. If I'm not sure exactly how the input amount is validated before reaching this class, toEqual would be a safe bet.

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 3 above)
Comment (mihai12) [reply coding-challenge/tst/bank/bank-account.test.ts:44]: I thought to also be sure that the message is the same and as far as I know toThrow does not check the message and neither the type.

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 4 above)
Comment (mihai12) [reply coding-challenge/tst/bank/bank-account.test.ts:80]: I think toEqual is better in this case, because it is possible that the user will input floating-numbers

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 5 above)
Comment (mihai12) [reply coding-challenge/tst/bank/bank-account.test.ts:104]: Same as above :)

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 6 above)
Comment (mihai12) [reply coding-challenge/tst/bank/bank-account.test.ts:55]: Yes, my fault.

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 7 above)
Comment (mihai12) [reply coding-challenge/tst/bank/bank-account.test.ts:131]: Same as above: 
1. I should have not used test methods in a unit test; 
2. Because the user could use floating-point numbers and toBe does not work well with them, I've changed toBe to toEqual everywhere.

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 8 above)
Comment (mihai12) [reply coding-challenge/tst/bank/bank-account.test.ts:114]: Indeed, it does deserve a const.

File: coding-challenge/tst/bank/bank-account.test.ts (hunk 9 above)
Comment (mihai12) [reply coding-challenge/tst/bank/bank-account.test.ts:123]: Indeed I should have added the withdraw function too, and use different error messages:

I should have added:

```
jest.spyOn(BankAccount.prototype, 'withdraw').mockImplementationOnce(() => {
    throw new Error('Insufficient funds!');
});
```
And for the deposit change the error to: throw new Error('Deposit failed!');
```
expect(() => {
        bankAccount.transfer(transferAmount, destinationAccount);
    }).toThrowError('Insufficient funds!');
```

Reset mocks and carry on with the deposit action:

```
   jest.clearAllMocks();
    jest.spyOn(BankAccount.prototype, 'withdraw').mockImplementationOnce(() => {
        bankAccount['balance'] -= transferAmount;
    });
    jest.spyOn(destinationAccount, 'deposit').mockImplementationOnce(() => {
        throw new Error('Deposit failed!');
    });

    expect(() => {
        bankAccount.transfer(transferAmount, destinationAccount);
    }).toThrowError('Deposit failed!');
```

File: coding-challenge/src/bank/bank-account.ts (hunk 10 above)
Comment (mihai12) [reply coding-challenge/src/bank/bank-account.ts:33]: I should have implemented proper error handling. A try block to wrap both the withdraw and the deposit functions and a catch with an error message.

File: coding-challenge/src/bank/validation.ts (hunk 11 above)
Comment (mihai12) [reply coding-challenge/src/bank/validation.ts:4]: Please see answer on the first question.

**Comment by adrianmota-sag** [comment]
As per Jests documentation: 

- toBe and toEqual are equivalent for numbers, that's why I asked why you changed the code; if you want to be as accurate as possible about testing floating point numbers you have to use .toBeCloseTo();
- toThrow also checks specific messages, actually, toThrow and toThrowError are aliases

File: coding-challenge/src/bank/bank-account.ts (hunk 10 above)
Comment (razvan-zavalichi) [reply coding-challenge/src/bank/bank-account.ts:33]: If I have 100$ and I want to transfer 50$ to you, my balance will be down to 50$ even if the withdraw method failed.
So, If the withdraw method fails we should do a rollback.
